### ISSUE TITLE
Database Encryption Interface

### ISSUE THREAD

**Issue by jjeffryes** [OpenBazaar/openbazaar-desktop#89]
When the client starts a local server (this should mean the node server and client were installed together through an installer file), the first time the node server runs it will need a password for the database encryption. 

If the user chooses not to store the password, they will also need to enter it every time the node server starts.

This will not be necessary for remote node servers, or any node server that is started with the command line. When  node server is started through the command line, the password will be handled through the command line at that time.

## Functional Requirements

- [ ] When the client starts, if the client is connecting to the bundled local server and if no encryption password is saved in the client, an interface should be presented to enter an encryption password. 
- [ ] The encryption password interface should be the same if it is a new password or an existing password (the client does not have data on whether this is the first time the server is running). 
- [ ] The interface for the encryption password should include an option to save the password, and automatically send it any time that server is started. 
    - Since the client does not know if the server is starting for the first time, the option to save the password would always be available, even if this password has been entered previously.
- [ ] A warning should be included that saving the password is not the most secure option.
- [ ] If the option to save the encryption password is activated, the encryption password should be saved by the client and automatically send with the startup command any time the bundled local server is started.
- [ ] The user should be required to back up their password (it could be exported as a text file). 
- [ ] The user should be informed that if they lose this password, they will lose all transactions and never be able to recover them. This should be very strongly emphasized.
- [ ] When the password is sent with the startup command, if it fails an error should be shown.
- [ ] If the startup command fails, the user should be able to submit a corrected password, or connect to a different server (this may mean the password interface should be part of the connection modal, or there is a way to open the connection modal from the encryption password interface). 

## Data

- [ ] When the command to start the server is sentif the startup fails an error code needs to be returned.
- [ ] The command to start the server should accept a parameter for an encryption password.
- [ ] If no encryption password exists, the server should accept a parameter for a new encryption password.
- [ ] When the command to start the server is sent with a new password, an error code needs to be returned if the new password cannot be set.
- [ ] When the command to start the server is sent with an existing password, an error code needs to be returned if the password is rejected.

**Comment by cpacia** [comment]
> The encryption password interface should be the same if it is a new password or an existing password (the client does not have data on whether this is the first time the server is running).

I don't know if that's the best way to handle it. At minimum the client will need to know if this is the first time the server is running so that it can load onboarding. 

My assumption is the "create password" dialog would look like a standard create pw dialog where the user enters the pw and then has to "confirm" it by typing it a second time in another text field. We'll have to figure out how we're going to detect first startup.

**Comment by morebrownies** [comment]
Ok I'll be working on the on-boarding step today, which will account for the following:

1. Set a password
2. Confirm the password (re-enter it)
3. Include a checkbox to save the password in the client (warn the user this is less secure approach)
4. Provide a way to backup the set password (a button to export a txt file)
5. Include instructions to explain what the password is used for

**Comment by rmisio** [comment]
It sounds like you two (Josh and Chris) may be saying the same thing. If the dialog is the same whether it's a new password creation or the entering of non-saved creds, then it simplifies the workflow.

- If is bundled app and user server config is set to connect to the Default (bundled) server:
   - If we have stored creds, attempt to launch the bundled server with the stored creds.
        - If server returns error code for failed auth, launch password dialog
    - if we *do not* have stored creds, launch password dialog

Onboarding has nothing to do with this. That the client already knows how to detect based on the settings and profile APIs. But, that is beyond this, because you already need a running and authenticated server for that part.

I think this is a good case for keeping the creds out of onboarding and in a seperate decoupled dialog. This way we would only need one singular database encryption creds dialog that is simply shown anytime we launch the bundled server and it exits with a BAD_CREDS error code.

**Comment by jjeffryes** [comment]
I agree that the encryption password dialog should stay separate from onboarding, since onboarding doesn't get launched until after the server is started, which is after the password needs to be entered (I assume). 

@cpacia I agree it would be ideal to have a slightly different experience when the password is set vs. entered, but how will the client know if the server has a password or not? Doesn't it need to send the password with the start command?

If instead the sequence is "start server, sever sends message it needs the password (can this be in the response from the start command, or does it need to be an API call?), then client shows the password interface" then we have more flexibility.

**Comment by morebrownies** [comment]
Here's my first pass at integrating the password step into on-boarding: 

https://invis.io/MN8WX2BU2#/196013031_Openbazaar-2-0-Splash-Screen

Thoughts? @jjeffryes @cpacia @rmisio

**Comment by jjeffryes** [comment]
@morebrownies see comments above. I don't think it can work this way.

My understanding is the sequence of events is:

1. The client starts up. If it was installed with a bundled server, it checks to see if it has a saved encryption password.
2. If it does, start the server with that password.
3. If it does not, it shows the password entry screen. 
4. After the server gets the password and starts, the client calls the server to see if the user has saved data.
5. If the user does not have saved data, Onboarding is shown.
6. If the user does have saved data, Onboarding is not shown. 

With the added wrinkle above that if we can figure out a way to tell the client if this is the first time the server is being started, the password entry UX will be a little different.

**Comment by morebrownies** [comment]
We really need some user flows put together for this section. @jjeffryes you seem to have a good understanding of what all of the possible scenarios are and where the possible edge cases are. 

Would you mind taking a stab at drawing them out to help get us all on the same page?

**Comment by jjeffryes** [comment]
@morebrownies will do.

**Comment by cpacia** [comment]
> Provide a way to backup the set password (a button to export a txt file)

Personally I feel like this is unnecessary as almost all computer users know how to copy and paste. It is also a major break from every other service/app that asks for a password as I'm not aware of any that have an export button.

**Comment by cpacia** [comment]
@jjeffryes if it's checking local storage for the password at startup, couldn't it also check for a boolean if the password had been set yet (meaning check if this is the very first startup)? 

If so then you could so the "set password" modal instead of the "enter password" modal.

**Comment by rmisio** [comment]
> if it's checking local storage for the password at startup, couldn't it also check for a boolean if the password had been set yet (meaning check if this is the very first startup)?

Not sure about that.. the solution is a bit brittle. Local storage can be cleared. A server could have it's data wiped. The user could use a different instance of the client to connect to the server.

Whether the password has been set is really a property of the server.

**Comment by rmisio** [comment]
I can't think of a way, but I'll throw it out there anyway... is there not a way we could simplify and somehow make the modal generic enough to support both cases where it's either create or re-enter password.

Looking at Mike's mockups, the only field that wouldn't work for both scenarios is the re-enter password field.

**Comment by jjeffryes** [comment]
I think exporting a text file is safer than copy and paste, because it requires the user actually does something.

If we just put text on the screen saying it's a really good idea to write the password down or copy and paste it, 99% of users will ignore that text.

Everyone knows there is always a "recover password" button somewhere, so they'll just assume that's the case and ignore the warning.

And then they'll forget their password and lose everything forever.

**Comment by morebrownies** [comment]
> Provide a way to backup the set password (a button to export a txt file)

> Personally I feel like this is unnecessary as almost all computer users know how to copy and paste. It is also a major break from every other service/app that asks for a password as I'm not aware of any that have an export button.

> Also seems to add to the cognitive load.

Service/apps have ways to reset your password, where as OpenBazaar does not. We need to be overly cautious here and make it extremly clear that the password can not be recovered while also providing a way to back up the password easily.

If not, we run the risk of many people bricking their entire store just because we want to avoid adding cognitive load.

**Comment by cpacia** [comment]
If the password is sitting in a file on disk in the clear it's not doing any good.

**Comment by cpacia** [comment]
> Looking at Mike's mockups, the only field that wouldn't work for both scenarios is the re-enter password field.

That really is a needed field. The error rate goes up a lot if it's not included.

**Comment by rmisio** [comment]
Yeah, I wasn't suggesting eliminating that field. I was just brainstorming if there are some ways to present that form (including that field) in a more generic way to avoid the main problem we have... which is: The client has no idea whether a password has ever been set. Ideally, this is something the server could communicate to the client. If the client stores a boolean itself for this, there are definitely edge cases where the client will not have the right value. Maybe we could live with those edge cases?

**Comment by cpacia** [comment]
It could possible check if the repo exists. After all that's what the server does to see if it needs to initialize everything.

```
if !repoExists {
     goto onboarding;
}
```

**Comment by rmisio** [comment]
oh man, lol. The client already has an algorythm and code in place to know whether onboarding is needed and it's based off of whether the profile and settings have been set, which are done via onboarding.

The client needs to know whether the password has been set, which is a different thing.

> It could possible check if the repo exists.

Not sure I understand. I'm not aware of the client doing anything with a server repo. I think that's already in place as a part of the unpacking process. The client just runs the server 'go run openbazaard.go start' via a child process.

Ideally, the server could start without the creds and then give us an API of some sort to know whether the password has been set or not.

Anyhow.. back to the option of the client keeping a flag to know whether the passwords have been set. Here are some edge cases:

*Client thinks the password is NOT set when it actually is*

1.) Client initially has the flag set as false.
2.) User creates creds and happily uses app for a while.
3.) User somehow clears their local storage.
4.) Next time they log-in, they are re-shown the Create password form. If they enter their previous creds, they'll be fine. If not, then I guess we'll get the failed auth error code and we could show them the other flavor of the form (the one where you already have creds and just need to enter them) to re-enter creds.

Another edge case:

*Client thinks the password IS set when it actually is NOT*

1.) Client initially has the flag set as false.
2.) User creates creds and happily uses app for a while.
3.) User wipes their server.
4.) Next time they log-in, they are shown the standard dialog to enter creds (not the create new creds form). So, they won't have a chance to confirm the password.

Another edge case, is very edgy, but it's if they somehow exported this bundled server to another server and then wanted to use a different client to connect to it. I think the first edge case from above would be in play.

So, I don't think the first edge case is so bad. The second one is mildly bad, but how likely is it that someone will wipe their bundled server (and then they'll have to have a typo in the password they create)?

@jjeffryes @morebrownies Do we live with the edge cases? Or do we push for the server to expose an API to let us know more what's going on?

**Comment by rmisio** [comment]
Sigh, I was hoping we'd simplify the connection process and we've taken a U turn from that path :(

**Comment by jjeffryes** [comment]
Let's talk this out on a call, I think that would be more efficient.

**Comment by rmisio** [comment]
Really, we're only at 21 comments (now 22) on this thread? Can't we hit 30 or 40 first?

**Comment by jjeffryes** [comment]
![encryption_password_flow](https://cloud.githubusercontent.com/assets/1584275/20370927/41acf9c2-ac2f-11e6-8c46-71a38c318d19.png)
